internal/controller/services/gateway/gateway_controller_actions.go

```diff
@@ -21,136 +21,308 @@ import (
 		"certificateType", getCertificateType(gatewayConfig))
 
 	return nil
 }
 
-// createDestinationRule creates a DestinationRule for TLS configuration using embedded YAML template.
-// This function uses embedded resources for efficient template management.
-func createDestinationRule(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
-	// Input validation
-	if rr == nil {
-		return errors.New("reconciliation request cannot be nil")
+// Check authentication mode and deploy auth proxy (secret + service + deployment) + OAuth client (if integrated mode) + HTTPRoute + DestinationRule.
+func createKubeAuthProxyInfrastructure(ctx context.Context, rr *odhtypes.ReconciliationRequest) error {
+	l := logf.FromContext(ctx).WithName("createAuthProxy")
+
+	gatewayConfig, ok := rr.Instance.(*serviceApi.GatewayConfig)
+	if !ok {
+		return errors.New("instance is not of type *services.GatewayConfig")
 	}
 
-	l := logf.FromContext(ctx).WithName("createDestinationRule")
-	l.V(1).Info("Creating DestinationRule for TLS configuration")
+	l.V(1).Info("creating auth proxy for gateway", "gateway", gatewayConfig.Name)
 
-	yamlContent, err := gatewayResources.ReadFile("resources/destinationrule-tls.yaml")
+	authMode, err := cluster.GetClusterAuthenticationMode(ctx, rr.Client)
 	if err != nil {
-		return fmt.Errorf("failed to read DestinationRule template: %w", err)
+		return fmt.Errorf("failed to detect cluster authentication mode: %w", err)
+	}
+	l.V(1).Info("detected cluster authentication mode", "mode", authMode)
+
+	// Reset the condition to false
+	rr.Conditions.MarkFalse(ReadyConditionType)
+
+	kubeAuthProxyDeploymentTemplates := odhtypes.TemplateInfo{
+		FS:   gatewayResources,
+		Path: kubeAuthProxyDeploymentOauthTemplate,
+	}
+
+	var oidcConfig *serviceApi.OIDCConfig
+	switch authMode {
+	case cluster.AuthModeOIDC:
+		if gatewayConfig.Spec.OIDC == nil {
+			rr.Conditions.MarkFalse(
+				ReadyConditionType,
+				conditions.WithReason(status.NotReadyReason),
+				conditions.WithMessage(status.AuthProxyOIDCModeWithoutConfigMessage),
+			)
+			return nil // TODO: is this logic correct? no oidc but to use oidc and not error out.
+		}
+		oidcConfig = gatewayConfig.Spec.OIDC
+		l.V(1).Info("configuring "+KubeAuthProxyName+" for external OIDC",
+			"issuerURL", oidcConfig.IssuerURL,
+			"clientID", oidcConfig.ClientID,
+			"secretRef", oidcConfig.ClientSecretRef.Name)
+		kubeAuthProxyDeploymentTemplates = odhtypes.TemplateInfo{
+			FS:   gatewayResources,
+			Path: kubeAuthProxyDeploymentOidcTemplate,
+		}
+	case cluster.AuthModeIntegratedOAuth: // default mode.
+		l.V(1).Info("configuring " + KubeAuthProxyName + " for OpenShift OAuth")
+
+	case cluster.AuthModeNone:
+		rr.Conditions.MarkTrue(
+			ReadyConditionType,
+			conditions.WithReason(status.ReadyReason), // TODO: is this logic correct? user do not want it, we should mark it as ready.
+			conditions.WithMessage(status.AuthProxyExternalAuthNoDeploymentMessage),
+		)
+		return nil
```

Comment:
I dont 3rd case is correct, it has to be either ODIC or OAUTH.
cc: @jctanner for more